File: src/main/java/org/micoli/php/ui/components/tasks/tree/TreeIterator.java
package org.micoli.php.ui.components.tasks.tree;

import com.intellij.ui.treeStructure.Tree;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.TreePath;

public class TreeIterator {
    @FunctionalInterface
    public interface LeafProcessor {
        void process(DefaultMutableTreeNode node, boolean isLeaf, TreePath path, int level, int index);
    }

    private final LeafProcessor leafProcessor;
    private int index;

    private TreeIterator(LeafProcessor leafProcessor) {
        this.leafProcessor = leafProcessor;
        this.index = -1;
    }

    public static void forEach(Tree tree, LeafProcessor leafProcessor) {
        new TreeIterator(leafProcessor)
                .forEachRecursive((DefaultMutableTreeNode) tree.getModel().getRoot(), 1);
    }

    private void forEachRecursive(DefaultMutableTreeNode node, int level) {
        index++;
        TreePath path = new TreePath(node.getPath());
        if (node.isLeaf()) {
            leafProcessor.process(node, true, path, level, index);
            return;
        }
        leafProcessor.process(node, false, path, level, index);
        for (int nodeIndex = 0; nodeIndex < node.getChildCount(); nodeIndex++) {
            forEachRecursive((DefaultMutableTreeNode) node.getChildAt(nodeIndex), level + 1);
        }
    }
}

File: src/main/kotlin/org/micoli/php/ui/components/tasks/tree/TreeIterator.kt
package org.micoli.php.ui.components.tasks.tree

import com.intellij.ui.treeStructure.Tree
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.TreePath

class TreeIterator private constructor(private val leafProcessor: LeafProcessor) {
    fun interface LeafProcessor {
        fun process(node: DefaultMutableTreeNode?, isLeaf: Boolean, path: TreePath?, level: Int, index: Int)
    }

    private var index: Int = -1

    private fun forEachRecursive(node: DefaultMutableTreeNode, level: Int) {
        index++
        val path = TreePath(node.path)
        if (node.isLeaf) {
            leafProcessor.process(node, true, path, level, index)
            return
        }
        leafProcessor.process(node, false, path, level, index)
        for (nodeIndex in 0..<node.childCount) {
            forEachRecursive((node.getChildAt(nodeIndex) as DefaultMutableTreeNode?)!!, level + 1)
        }
    }

    companion object {
        fun forEach(tree: Tree, leafProcessor: LeafProcessor) {
            TreeIterator(leafProcessor).forEachRecursive((tree.model.root as DefaultMutableTreeNode?)!!, 1)
        }
    }
}
